src/main/java/org/apache/ibatis/builder/annotation/ProviderSqlSource.java

```diff
@@ -116,33 +118,45 @@ private SqlSource createSqlSource(Object parameterObject) {
+        if (bindParameterCount == 1 && providerMethodParameterTypes[0] == Map.class) {
+          sql = invokeProviderMethod(extractProviderMethodArguments(parameterObject));
+        } else {
+          @SuppressWarnings("unchecked")
+          Map<String, Object> params = (Map<String, Object>) parameterObject;
+          sql = invokeProviderMethod(extractProviderMethodArguments(params, providerMethodArgumentNames));
+        }
+      } else if (providerMethodParameterTypes.length == 0) {
         sql = invokeProviderMethod();
-      } else if (bindParameterCount == 0) {
-        sql = invokeProviderMethod(providerContext);
-      } else if (bindParameterCount == 1
-           && (parameterObject == null || providerMethodParameterTypes[providerContextIndex == null || providerContextIndex == 1 ? 0 : 1].isAssignableFrom(parameterObject.getClass()))) {
+      } else if (providerMethodParameterTypes.length == 1) {
+        if (providerContext == null) {
+          sql = invokeProviderMethod(parameterObject);
+        } else {
+          sql = invokeProviderMethod(providerContext);
+        }
+      } else if (providerMethodParameterTypes.length == 2) {
         sql = invokeProviderMethod(extractProviderMethodArguments(parameterObject));
-      } else if (parameterObject instanceof Map) {
-        @SuppressWarnings("unchecked")
-        Map<String, Object> params = (Map<String, Object>) parameterObject;
-        sql = invokeProviderMethod(extractProviderMethodArguments(params, providerMethodArgumentNames));
       } else {
-        throw new BuilderException("Error invoking SqlProvider method ("
-                + providerType.getName() + "." + providerMethod.getName()
-                + "). Cannot invoke a method that holds "
-                + (bindParameterCount == 1 ? "named argument(@Param)" : "multiple arguments")
-                + " using a specifying parameterObject. In this case, please specify a 'java.util.Map' object.");
+        throw new BuilderException("Cannot invoke SqlProvider method '" + providerMethod
+          + "' with specify parameter '" + (parameterObject == null ? null : parameterObject.getClass())
+          + "' because SqlProvider method arguments for '" + mapperMethod + "' is an invalid combination.");
       }
       Class<?> parameterType = parameterObject == null ? Object.class : parameterObject.getClass();
       return languageDriver.createSqlSource(configuration, sql, parameterType);
     } catch (BuilderException e) {
       throw e;
     } catch (Exception e) {
-      throw new BuilderException("Error invoking SqlProvider method ("
-          + providerType.getName() + "." + providerMethod.getName()
-          + ").  Cause: " + e, e);
+      throw new BuilderException("Error invoking SqlProvider method '" + providerMethod
+          + "' with specify parameter '" + (parameterObject == null ? null : parameterObject.getClass()) + "'.  Cause: " + extractRootCause(e), e);
```

Comment:
I think it's not necessary to include the cause in the message.
It will show up in the stack trace anyway.

Reply:
I agree with your opinion. However, Some other classes have included exception debugging information on the cause exception in the same way, so I think that it should be fixed separately from this issue. WDYT?

Reply:
I've searched with "Cause:", it matched at 72 point in 34 files on src/main.

Reply:
No problem. :)